File: +bsv/fetchConnectivityData.m
function [combinedProjection, combinedInjectionInfo, individualProjections, experimentRegionInfo] = fetchConnectivityData(experimentIDs, saveLocation, fileName, ...
    normalizationMethod, subtractOtherHemisphere, groupingMethod, allenAtlasPath, loadAll, inputRegions, regionGroups, exportMetadata, reload, atlasType, atlasResolution)

if nargin < 6 || isempty(groupingMethod) || nargin < 7 || isempty(allenAtlasPath) % group experiments by brain region (groupingMethod = 'region') or not
    groupingMethod = 'NaN';
%else
    % st = loadStructureTree([allenAtlasPath, filesep, 'structure_tree_safe_2017.csv']); % a table of what all the labels mean
    % I will need this to implement some selection of regions by hierarchy.
    % 
end

% Handle optional input regions and region groups
if nargin < 9 || isempty(inputRegions)
    inputRegions = [];
end
if nargin < 10 || isempty(regionGroups)
    regionGroups = [];
end
if nargin < 11 || isempty(exportMetadata)
    exportMetadata = true;
end
if nargin < 12 || isempty(reload)
    reload = false;
end
if nargin < 13 || isempty(atlasType)
    atlasType = 'allen'; % Default to Allen atlas
end
if nargin < 14 || isempty(atlasResolution)
    atlasResolution = 10; % Default to 10um resolution
end

%% Construct atlas filenames based on type and resolution
switch lower(atlasType)
    case 'allen'
        if atlasResolution == 10
            annotationFile = 'annotation_volume_10um_by_index.npy';
            structureFile = 'structure_tree_safe_2017.csv';
        elseif atlasResolution == 20
            annotationFile = 'annotation_volume_v2_20um_by_index.npy';
            structureFile = 'UnifiedAtlas_Label_ontology_v2.csv';
        else
            error('Unsupported Allen atlas resolution: %d. Supported resolutions are 10 and 20 um.', atlasResolution);
        end
    otherwise
        % For custom atlases, construct filename from type and resolution
        annotationFile = sprintf('%s_annotation_%dum.npy', atlasType, atlasResolution);
        structureFile = sprintf('%s_structure_tree.csv', atlasType);
end

%% fetch data

projectionGridSize = [132, 80, 114];

% Create main save directory if it doesn't exist
if ~exist(saveLocation, 'dir')
    try
        mkdir(saveLocation);
    catch ME
        warning('Could not create directory %s: %s\nUsing current directory instead.', saveLocation, ME.message);
        saveLocation = pwd;
    end
end

% filepaths - include atlas info in filename
atlasStr = sprintf('_%s_%dum', atlasType, atlasResolution);
filePath_imgs = [saveLocation, filesep, fileName, '_', normalizationMethod, '_sub', num2str(subtractOtherHemisphere), atlasStr, '.mat'];
filePath_injectionSummary = [saveLocation, filesep, fileName, '_injectionSummary', atlasStr, '.mat'];

% load summary from the Allen
currentFileLocation =  which(mfilename('fullpath'));
currentFileLocation_Pa = fileparts(fileparts(currentFileLocation)); % twp directories up from current file location
allenAtlasProjection_info = readtable(fullfile(currentFileLocation_Pa, 'docs', 'allenAtlasProjection_info.csv'), 'VariableNamingRule','modify');

% intialize arrays
nExpIDs = size(experimentIDs, 2); % added by Matteo
primaryStructure_AP = zeros(nExpIDs, 1);
primaryStructure_DV = zeros(nExpIDs, 1);
primaryStructure_ML = zeros(nExpIDs, 1);
primaryStructure_ID = zeros(nExpIDs, 1);
primaryStructure_abbreviation = cell(nExpIDs, 1);



    combinedInjectionInfo = table;
    
    % display progress
    disp(['Loading ', num2str(nExpIDs), ' experiments...']);
    progressBarHandle = waitbar(0, 'Loading experiments...');

    for iExpID = 1:nExpIDs
        waitbar(iExpID/nExpIDs, progressBarHandle, sprintf('Loading experiment %d of %d', iExpID, nExpIDs));

        currExpID = experimentIDs(iExpID);
        % create dir if it doesn't exist
        saveDir = [saveLocation, filesep, num2str(currExpID)];
        if ~exist(saveDir, 'dir')
            try
                [status, msg] = mkdir(saveDir);
                if ~status
                    error('Failed to create directory %s: %s', saveDir, msg);
                end
            catch ME
                warning('Could not create experiment directory %s: %s\nSkipping download for experiment %d.', saveDir, ME.message, currExpID);
                continue;
            end
        end

        % summary (structure.ionizes)
        summaryFilePath = fullfile(saveLocation, num2str(currExpID), 'injectionSummary_all.mat');
        if ~exist(summaryFilePath, 'file')
            status = bsv.fetchConnectivitySummary(currExpID, [saveLocation, filesep, num2str(currExpID)]);
            if ~status
                continue;
            end
        end
        load(summaryFilePath, 'injectionInfo'); % MC specified what to load

        % put all strcuture.ionizes summary into one table
        % Initialize combinedInjectionInfo if it doesn't exist
        if iExpID == 1
            combinedInjectionInfo = struct();
            fieldNames = fieldnames(injectionInfo);
        end

        for iFieldName = 1:numel(fieldNames)
            % Ensure the field exists in combinedInjectionInfo
            if ~isfield(combinedInjectionInfo, fieldNames{iFieldName})
                combinedInjectionInfo.(fieldNames{iFieldName}) = [];
            end

            % Concatenate the data
            if size(combinedInjectionInfo.(fieldNames{iFieldName}), 1) == 1
                combinedInjectionInfo.(fieldNames{iFieldName}) = [combinedInjectionInfo.(fieldNames{iFieldName}), [injectionInfo.(fieldNames{iFieldName})]];
            else
                combinedInjectionInfo.(fieldNames{iFieldName}) = [combinedInjectionInfo.(fieldNames{iFieldName}); [injectionInfo.(fieldNames{iFieldName})]];
            end
        end
        currInjectionInfo = struct2table(injectionInfo);
        primaryStructure_ID(iExpID) = allenAtlasProjection_info.structure_id(allenAtlasProjection_info.id == currExpID); % this should hold true 
        % in most cases. the best way would be to take into account group hierarchies but they
        % make no sense to me for instance VISp is at level 7, CP and GPe at level 6, SNr at level
        % 5 but conceptually they are at the same level to me...
        primaryStructure_abbreviation{iExpID} = allenAtlasProjection_info.structure_abbrev{allenAtlasProjection_info.id == currExpID};
        currInjectionCoordinates = str2num(allenAtlasProjection_info.injection_coordinates{allenAtlasProjection_info.id == currExpID});
        primaryStructure_AP(iExpID) = currInjectionCoordinates(1);
        primaryStructure_DV(iExpID) = currInjectionCoordinates(2);
        primaryStructure_ML(iExpID) = currInjectionCoordinates(3);
    end
    close(progressBarHandle);

    %% Display experiment summary
    % Filter for current experiments
    expInfo = allenAtlasProjection_info(ismember(allenAtlasProjection_info.id, experimentIDs), :);
    
    fprintf('\n📊 EXPERIMENT SUMMARY\n');
    fprintf('Total experiments loaded: %d\n', length(experimentIDs));
    
    % Mouse genotype distribution
    fprintf('\n🧬 Mouse genotype distribution:\n');
    genotypes = expInfo.transgenic_line;
    
    % Handle empty genotypes
    genotypes(cellfun(@isempty, genotypes)) = {'Wild-type'};
    genotypes(strcmp(genotypes, '""') | strcmp(genotypes, '')) = {'Wild-type'};
    
    [uniqueGenotypes, ~, genotypeIdx] = unique(genotypes);
    for i = 1:length(uniqueGenotypes)
        count = sum(genotypeIdx == i);
        fprintf('  • %s: %d experiments\n', uniqueGenotypes{i}, count);
    end
    
    % Brain region distribution
    fprintf('\n🧠 Brain region distribution:\n');
    regions = expInfo.structure_abbrev;
    [uniqueRegions, ~, regionIdx] = unique(regions);
    for i = 1:length(uniqueRegions)
        count = sum(regionIdx == i);
        fprintf('  • %s: %d experiments\n', uniqueRegions{i}, count);
    end
    fprintf('─────────────────────────────────\n\n');

    % % grouping method 
    if strcmp(groupingMethod, 'brainRegion')
        % because of hierarchy differences, this doesn't make much sense
        % currently, not implemented 
        warning('grouping method not implemented yet - skipping grouping. ')
        % [sorted_values, sort_indices] = st.acronym(st.id == combinedInjectionInfo.structure_id);
        % [groups, ~, groupID] = unique(primaryStructure_ID);
        % groupID_original_order = zeros(size(groupID));
        % groupID_original_order(sort_indices) = groupID;
    elseif strcmp(groupingMethod, 'AP') % group by AP value
        [sorted_values, sort_indices] = sort(primaryStructure_AP);
        [groups, ~, groupID] = unique(sorted_values);
        groupID_original_order = zeros(size(groupID));
        groupID_original_order(sort_indices) = groupID;
    elseif strcmp(groupingMethod, 'ML') % group by ML value
        [sorted_values, sort_indices] = sort(primaryStructure_ML);
        [groups, ~, groupID] = unique(sorted_values);
        groupID_original_order = zeros(size(groupID));
        groupID_original_order(sort_indices) = groupID;
    elseif strcmp(groupingMethod, 'DV') % group by DV value
        [sorted_values, sort_indices] = sort(primaryStructure_DV);
        [groups, ~, groupID] = unique(sorted_values);
        groupID_original_order = zeros(size(groupID));
        groupID_original_order(sort_indices) = groupID;
    elseif strcmp(groupingMethod, 'NaN') || isempty(groupingMethod) % no grouping
        groups = ones(size(experimentIDs, 2), 1);
        groupID_original_order = ones(size(experimentIDs, 2), 1);
    else % error
        warning('grouping method not recognized - skipping grouping. ')
        groups = ones(size(experimentIDs, 2), 1);
        groupID_original_order = ones(size(experimentIDs, 2), 1);
    end


    numberOfGroups = numel(unique(groups));
    
    % Override grouping if region-based grouping is requested
    if ~isempty(inputRegions) && ~isempty(regionGroups)
        fprintf('Setting up region-based grouping for %d input regions into %d groups\n', length(inputRegions), length(unique(regionGroups)));
        
        % Create region-to-group mapping
        uniqueRegionGroups = unique(regionGroups);
        nRegionGroups = length(uniqueRegionGroups);
        regionGroupsCellArray = cell(nRegionGroups, 1);
        
        for i = 1:nRegionGroups
            regionGroupsCellArray{i} = find(regionGroups == uniqueRegionGroups(i));
        end
        
        % Map each experiment to its region group
        regionBasedGroupID = zeros(nExpIDs, 1);
        unmatchedExperiments = {};
        for iExpID = 1:nExpIDs
            regionAbbrev = primaryStructure_abbreviation{iExpID};
            
            % Find which input region this experiment corresponds to
            regionIdx = 0;
            for iRegion = 1:length(inputRegions)
                if strcmp(regionAbbrev, inputRegions{iRegion}) || startsWith(regionAbbrev, inputRegions{iRegion})
                    regionIdx = iRegion;
                    break;
                end
            end
            
            % Find which region group this region belongs to
            if regionIdx > 0
                for iRegionGroup = 1:nRegionGroups
                    if any(regionGroupsCellArray{iRegionGroup} == regionIdx)
                        regionBasedGroupID(iExpID) = iRegionGroup;
                        break;
                    end
                end
            else
                % Log unmatched experiments
                unmatchedExperiments{end+1} = sprintf('Exp %d: %s', experimentIDs(iExpID), regionAbbrev);
            end
        end
        
        % Report unmatched experiments
        if ~isempty(unmatchedExperiments)
            fprintf('\nWarning: %d experiments did not match any input region:\n', length(unmatchedExperiments));
            for i = 1:min(10, length(unmatchedExperiments))
                fprintf('  %s\n', unmatchedExperiments{i});
            end
            if length(unmatchedExperiments) > 10
                fprintf('  ... and %d more\n', length(unmatchedExperiments) - 10);
            end
        end
        
        % Override the grouping variables
        numberOfGroups = nRegionGroups;
        groupID_original_order = regionBasedGroupID;
        
        fprintf('Region-based grouping: %d experiments mapped to %d groups\n', sum(regionBasedGroupID > 0), numberOfGroups);
    end
    
    combinedProjection = zeros([projectionGridSize, numberOfGroups]);

    if loadAll
        individualProjections = zeros([projectionGridSize, size(experimentIDs, 2)]);
    else
        individualProjections = '';
    end

    % get raw images
    disp('Getting raw images...');
    progressBarHandle = waitbar(0, 'Getting raw images...');
    for iExpID = 1:nExpIDs
        waitbar(iExpID/nExpIDs, progressBarHandle, sprintf('Getting raw image %d of %d', iExpID, nExpIDs));

        currGroup = groupID_original_order(iExpID);
        currExpID = experimentIDs(iExpID);
        
        % Skip experiments that don't belong to any group
        if currGroup == 0
            continue;
        end

        % raw data
        rawFilePath = fullfile(saveLocation, num2str(currExpID), 'density.raw');
        if ~exist(rawFilePath, 'file')
            % fetch and save raw data if it isn't already on disk
            status = bsv.fetchConnectivityImages(currExpID, fullfile(saveLocation, num2str(currExpID)));
            if ~status
                continue;
            end
        end
        
        % load raw data file
        fid = fopen(rawFilePath, 'r', 'l');
        experiment_projection = fread(fid, prod(projectionGridSize), 'float');
        fclose(fid);
        experiment_projection = reshape(experiment_projection, projectionGridSize);
        
        % normalize projection values
        if strcmp(normalizationMethod, 'injectionVolume') || strcmp(normalizationMethod, 'injectionIntensity')

            % Find indices for the current experiment and structure 997 (root)
            expIndices = combinedInjectionInfo.experimentID == currExpID;
            structureIndices = expIndices & combinedInjectionInfo.structure_id == 997;

            % Use hemisphere 3 (both hemispheres combined)
            hem3Indices = structureIndices & [combinedInjectionInfo.hemisphere_id] == 3;

            % Both methods normalize by projection_volume (mm^3).
            % The density.raw values are already fractional (0-1), so dividing
            % by injection volume makes experiments with different injection
            % sizes comparable.
            injVol = max(combinedInjectionInfo.projection_volume(hem3Indices));

            if isempty(injVol) || injVol == 0
                fprintf('Warning: Invalid injection volume for experiment %d, skipping normalization\n', currExpID);
                injVol = 1;
            end

            experiment_projection = experiment_projection / injVol;
        end

        if subtractOtherHemisphere
            experiment_projection_tmp = zeros(132, 80, 114);
            if injectionInfo.max_voxel_z <= 114 / 2 %left
                for iML = 1:57
                    experiment_projection_tmp(:, :, iML) = [experiment_projection(:, :, iML) - experiment_projection(:, :, 114-iML+1)];
                end
            elseif injectionInfo.max_voxel_z >= 114 / 2 %right
                for iML = 1:57
                    experiment_projection_tmp(:, :, 114-iML+1) = [experiment_projection(:, :, 114-iML+1) - experiment_projection(:, :, iML)];
                end
            end
            experiment_projection = experiment_projection_tmp;
        end
        if loadAll
            individualProjections(:, :, :, iExpID) = experiment_projection;
        end
        combinedProjection(:, :, :, currGroup) = combinedProjection(:, :, :, currGroup) + experiment_projection;


    end
    close(progressBarHandle);

    % normalize images to number of projections - also need to add
    % normalization to injection intensity / volume
    for iGroup = 1:numberOfGroups
        theseGroups = numel(find(groupID_original_order == iGroup));
        combinedProjection(:, :, :, iGroup) = combinedProjection(:, :, :, iGroup) ./ theseGroups;
    end

    % Create experiment region info structure
    experimentRegionInfo.experimentIDs = experimentIDs;
    experimentRegionInfo.primaryStructure_ID = primaryStructure_ID;
    experimentRegionInfo.primaryStructure_abbreviation = primaryStructure_abbreviation;
    experimentRegionInfo.primaryStructure_AP = primaryStructure_AP;
    experimentRegionInfo.primaryStructure_DV = primaryStructure_DV;
    experimentRegionInfo.primaryStructure_ML = primaryStructure_ML;
    
    % Export metadata to CSV if requested
    if exportMetadata
        % Use default filename if none provided
        csvFileName_base = fileName;
        if isempty(csvFileName_base) || strcmp(csvFileName_base, '')
            csvFileName_base = 'connectivity_data';
            fprintf('No filename provided for metadata export, using default: %s\n', csvFileName_base);
        end
        fprintf('Exporting experiment metadata to CSV...\n');
        
        % Create metadata table
        metadataTable = table();
        
        % Basic experiment info
        metadataTable.ExperimentID = experimentIDs(:);
        
        % Get experiment info from Allen Atlas data
        expInfo = allenAtlasProjection_info(ismember(allenAtlasProjection_info.id, experimentIDs), :);
        
        % Ensure the order matches experimentIDs
        [~, idx] = ismember(experimentIDs, expInfo.id);
        expInfo = expInfo(idx, :);
        
        % Mouse line/genotype
        metadataTable.MouseLine = expInfo.transgenic_line;
        % Clean up empty genotypes
        emptyIdx = cellfun(@isempty, metadataTable.MouseLine) | strcmp(metadataTable.MouseLine, '""') | strcmp(metadataTable.MouseLine, '');
        metadataTable.MouseLine(emptyIdx) = {'Wild-type'};
        
        % Injection region
        metadataTable.InjectionRegion = primaryStructure_abbreviation(:);
        metadataTable.InjectionRegionID = primaryStructure_ID(:);
        
        % Injection coordinates
        metadataTable.InjectionAP = primaryStructure_AP(:);
        metadataTable.InjectionDV = primaryStructure_DV(:);
        metadataTable.InjectionML = primaryStructure_ML(:);
        
        % Get injection volume and intensity from combinedInjectionInfo
        % We need to extract the primary injection site data
        injectionVolumes = zeros(length(experimentIDs), 1);
        injectionIntensities = zeros(length(experimentIDs), 1);
        
        for iExp = 1:length(experimentIDs)
            expID = experimentIDs(iExp);
            
            % Find injection data for this experiment (structure_id = 997, hemisphere_id = 3)
            expIdx = combinedInjectionInfo.experimentID == expID;
            structIdx = combinedInjectionInfo.structure_id == 997;
            hemIdx = combinedInjectionInfo.hemisphere_id == 3;
            injIdx = expIdx & structIdx & hemIdx;
            
            if any(injIdx)
                % Handle multiple entries by taking the first/max value
                volumes = combinedInjectionInfo.projection_volume(injIdx);
                intensities = combinedInjectionInfo.sum_pixel_intensity(injIdx);
                
                if length(volumes) > 1
                    fprintf('Warning: Multiple injection entries found for experiment %d, using max values\n', expID);
                    injectionVolumes(iExp) = max(volumes);
                    injectionIntensities(iExp) = max(intensities);
                else
                    injectionVolumes(iExp) = volumes;
                    injectionIntensities(iExp) = intensities;
                end
            end
        end
        
        metadataTable.InjectionVolume = injectionVolumes;
        metadataTable.InjectionIntensity = injectionIntensities;
        
        % Sex information if available
        if ismember('sex', expInfo.Properties.VariableNames)
            metadataTable.Sex = expInfo.sex;
        end
        
        % Age information if available
        if ismember('age', expInfo.Properties.VariableNames)
            metadataTable.Age = expInfo.age;
        end
        
        % Save CSV file
        csvFileName = [saveLocation, filesep, csvFileName_base, '_metadata.csv'];
        writetable(metadataTable, csvFileName);
        fprintf('Metadata exported to: %s\n', csvFileName);
    end
disp('Data processing complete.');
end

File: bsv/fetch_connectivity_data.py
import os
import json
import numpy as np
import pandas as pd

from .fetch_connectivity_summary import fetch_connectivity_summary, load_injection_summary
from .fetch_connectivity_images import fetch_connectivity_images, PROJECTION_GRID_SIZE
from .atlas_utils import load_projection_info


def fetch_connectivity_data(experiment_ids, save_location, file_name,
                            normalization_method, subtract_other_hemisphere,
                            grouping_method='', allen_atlas_path='',
                            load_all=False, input_regions=None, region_groups=None,
                            export_metadata=True, reload=False,
                            atlas_type='allen', atlas_resolution=10):
    """Download and cache projection density maps for a set of experiments.

    Parameters
    ----------
    experiment_ids : list of int
        Experiment IDs returned by :func:`find_connectivity_experiments`.
    save_location : str
        Local directory for cached downloads.
    file_name : str
        Base name for the cached metadata CSV. ``''`` to skip caching.
    normalization_method : str
        ``'none'`` or ``'injectionIntensity'`` (divide by injection volume).
    subtract_other_hemisphere : bool
        Subtract the contralateral hemisphere signal.
    grouping_method : str, optional
        Group experiments by ``'AP'``, ``'ML'``, ``'DV'``, or ``''`` (no
        grouping).
    allen_atlas_path : str, optional
        Path to the Allen CCF atlas directory.
    load_all : bool, optional
        If True, also return individual (non-averaged) projection volumes.
    input_regions : list of str, optional
        Region acronyms for region-based grouping.
    region_groups : list of int, optional
        Group assignment per region (same length as *input_regions*).
    export_metadata : bool, optional
        Write a per-experiment metadata CSV.
    reload : bool, optional
        Re-download even if cached files exist.
    atlas_type : str, optional
        Atlas type (default ``'allen'``).
    atlas_resolution : int, optional
        Atlas resolution in micrometres (10 or 20).

    Returns
    -------
    combined_projection : numpy.ndarray
        Averaged projection density array (AP x DV x ML x groups).
    combined_injection_info : dict
        Aggregated injection metadata across experiments.
    individual_projections : numpy.ndarray or None
        Per-experiment volumes when *load_all* is True, else None.
    experiment_region_info : dict
        Per-experiment region metadata.
    """
    if not grouping_method:
        grouping_method = 'NaN'
    if input_regions is None:
        input_regions = []
    if region_groups is None:
        region_groups = []

    os.makedirs(save_location, exist_ok=True)

    # Load projection info CSV
    projection_info = load_projection_info()

    n_exp = len(experiment_ids)
    primary_structure_AP = np.zeros(n_exp)
    primary_structure_DV = np.zeros(n_exp)
    primary_structure_ML = np.zeros(n_exp)
    primary_structure_ID = np.zeros(n_exp, dtype=int)
    primary_structure_abbreviation = [''] * n_exp

    # Collect all injection info
    combined_injection_info = {}

    print(f'Loading {n_exp} experiments...')

    for i, exp_id in enumerate(experiment_ids):
        save_dir = os.path.join(save_location, str(exp_id))
        os.makedirs(save_dir, exist_ok=True)

        # Fetch summary if needed
        summary_path = os.path.join(save_dir, 'injectionSummary_all.json')
        if not os.path.exists(summary_path):
            status = fetch_connectivity_summary(exp_id, save_dir)
            if not status:
                continue

        injection_info = load_injection_summary(save_dir)

        # Accumulate injection info
        for entry in injection_info:
            for key, val in entry.items():
                if key not in combined_injection_info:
                    combined_injection_info[key] = []
                combined_injection_info[key].append(val)

        # Get metadata from projection info CSV
        row = projection_info[projection_info['id'] == exp_id]
        if len(row) > 0:
            row = row.iloc[0]
            primary_structure_ID[i] = row['structure_id']
            primary_structure_abbreviation[i] = row['structure_abbrev']
            coords_str = str(row['injection_coordinates'])
            coords = [float(x) for x in coords_str.strip('[]').split(',')]
            if len(coords) >= 3:
                primary_structure_AP[i] = coords[0]
                primary_structure_DV[i] = coords[1]
                primary_structure_ML[i] = coords[2]

    # Print experiment summary
    exp_info = projection_info[projection_info['id'].isin(experiment_ids)]
    print(f'\nEXPERIMENT SUMMARY')
    print(f'Total experiments loaded: {n_exp}')

    print(f'\nMouse genotype distribution:')
    genotypes = exp_info['transgenic_line'].fillna('Wild-type').replace({'': 'Wild-type', '""': 'Wild-type'})
    for gt, count in genotypes.value_counts().items():
        print(f'  {gt}: {count} experiments')

    print(f'\nBrain region distribution:')
    for reg, count in exp_info['structure_abbrev'].value_counts().items():
        print(f'  {reg}: {count} experiments')

    # Grouping
    if grouping_method in ('AP', 'ML', 'DV'):
        coord_map = {'AP': primary_structure_AP, 'ML': primary_structure_ML, 'DV': primary_structure_DV}
        vals = coord_map[grouping_method]
        sorted_idx = np.argsort(vals)
        unique_vals = np.unique(vals[sorted_idx])
        group_id = np.zeros(n_exp, dtype=int)
        for j, v in enumerate(unique_vals):
            group_id[vals == v] = j + 1
        n_groups = len(unique_vals)
    elif grouping_method == 'NaN' or not grouping_method:
        group_id = np.ones(n_exp, dtype=int)
        n_groups = 1
    else:
        print(f'Warning: grouping method "{grouping_method}" not recognized - skipping grouping.')
        group_id = np.ones(n_exp, dtype=int)
        n_groups = 1

    # Region-based grouping override
    if input_regions and region_groups:
        unique_rg = sorted(set(region_groups))
        n_region_groups = len(unique_rg)
        rg_cell = {g: [j for j, rg in enumerate(region_groups) if rg == g] for g in unique_rg}

        region_based_group_id = np.zeros(n_exp, dtype=int)
        for i in range(n_exp):
            abbrev = primary_structure_abbreviation[i]
            region_idx = -1
            for j, reg in enumerate(input_regions):
                if abbrev == reg or abbrev.startswith(reg):
                    region_idx = j
                    break
            if region_idx >= 0:
                for g_idx, g in enumerate(unique_rg):
                    if region_idx in rg_cell[g]:
                        region_based_group_id[i] = g_idx + 1
                        break

        n_groups = n_region_groups
        group_id = region_based_group_id
        print(f'Region-based grouping: {np.sum(group_id > 0)} experiments mapped to {n_groups} groups')

    # Initialize combined projection
    combined_projection = np.zeros((*PROJECTION_GRID_SIZE, n_groups))
    individual_projections = np.zeros((*PROJECTION_GRID_SIZE, n_exp)) if load_all else None

    # Load raw images and accumulate
    print('Getting raw images...')
    for i, exp_id in enumerate(experiment_ids):
        curr_group = group_id[i]
        if curr_group == 0:
            continue

        raw_path = os.path.join(save_location, str(exp_id), 'density.raw')
        if not os.path.exists(raw_path):
            status = fetch_connectivity_images(exp_id, os.path.join(save_location, str(exp_id)))
            if not status:
                continue

        # Read density.raw - MATLAB reads column-major (Fortran order)
        experiment_projection = np.fromfile(raw_path, dtype='<f4').reshape(PROJECTION_GRID_SIZE, order='F')

        # Normalize by injection volume (mm^3).
        # The density.raw values are already fractional (0-1), so dividing
        # by injection volume makes experiments with different injection
        # sizes comparable.
        if normalization_method in ('injectionVolume', 'injectionIntensity'):
            exp_indices = np.array(combined_injection_info.get('experimentID', [])) == exp_id
            struct_indices = exp_indices & (np.array(combined_injection_info.get('structure_id', [])) == 997)

            hem_ids = np.array(combined_injection_info.get('hemisphere_id', []))
            hem3 = struct_indices & (hem_ids == 3)

            vol_arr = np.array(combined_injection_info.get('projection_volume', []), dtype=float)
            inj_vol = vol_arr[hem3].max() if hem3.any() else 0

            if inj_vol == 0:
                print(f'Warning: Invalid injection volume for experiment {exp_id}, skipping normalization')
                inj_vol = 1

            experiment_projection = experiment_projection / inj_vol

        # Subtract other hemisphere
        if subtract_other_hemisphere:
            injection_info = load_injection_summary(os.path.join(save_location, str(exp_id)))
            max_voxel_z = injection_info[0].get('max_voxel_z', 57)
            tmp = np.zeros_like(experiment_projection)
            if max_voxel_z <= 57:  # left hemisphere
                for ml in range(57):
                    tmp[:, :, ml] = experiment_projection[:, :, ml] - experiment_projection[:, :, 113 - ml]
            else:  # right hemisphere
                for ml in range(57):
                    tmp[:, :, 113 - ml] = experiment_projection[:, :, 113 - ml] - experiment_projection[:, :, ml]
            experiment_projection = tmp

        if load_all:
            individual_projections[:, :, :, i] = experiment_projection
        combined_projection[:, :, :, curr_group - 1] += experiment_projection

    # Average by group count
    for g in range(n_groups):
        count = np.sum(group_id == g + 1)
        if count > 0:
            combined_projection[:, :, :, g] /= count

    # Build experiment region info
    experiment_region_info = {
        'experimentIDs': experiment_ids,
        'primaryStructure_ID': primary_structure_ID,
        'primaryStructure_abbreviation': primary_structure_abbreviation,
        'primaryStructure_AP': primary_structure_AP,
        'primaryStructure_DV': primary_structure_DV,
        'primaryStructure_ML': primary_structure_ML,
    }

    # Export metadata CSV
    if export_metadata:
        csv_base = file_name if file_name else 'connectivity_data'
        meta = pd.DataFrame()
        meta['ExperimentID'] = experiment_ids

        exp_info_ordered = projection_info[projection_info['id'].isin(experiment_ids)]
        idx_map = {eid: i for i, eid in enumerate(experiment_ids)}
        exp_info_ordered = exp_info_ordered.copy()
        exp_info_ordered['_sort'] = exp_info_ordered['id'].map(idx_map)
        exp_info_ordered = exp_info_ordered.sort_values('_sort').drop(columns='_sort')

        meta['MouseLine'] = exp_info_ordered['transgenic_line'].fillna('Wild-type').replace(
            {'': 'Wild-type', '""': 'Wild-type'}).values
        meta['InjectionRegion'] = primary_structure_abbreviation
        meta['InjectionRegionID'] = primary_structure_ID
        meta['InjectionAP'] = primary_structure_AP
        meta['InjectionDV'] = primary_structure_DV
        meta['InjectionML'] = primary_structure_ML

        # Injection volume/intensity from combined info
        inj_volumes = np.zeros(n_exp)
        inj_intensities = np.zeros(n_exp)
        exp_id_arr = np.array(combined_injection_info.get('experimentID', []))
        struct_id_arr = np.array(combined_injection_info.get('structure_id', []))
        hem_id_arr = np.array(combined_injection_info.get('hemisphere_id', []))
        pv_arr = np.array(combined_injection_info.get('projection_volume', []), dtype=float)
        spi_arr = np.array(combined_injection_info.get('sum_pixel_intensity', []), dtype=float)

        for j, eid in enumerate(experiment_ids):
            mask = (exp_id_arr == eid) & (struct_id_arr == 997) & (hem_id_arr == 3)
            if mask.any():
                inj_volumes[j] = pv_arr[mask].max()
                inj_intensities[j] = spi_arr[mask].max()

        meta['InjectionVolume'] = inj_volumes
        meta['InjectionIntensity'] = inj_intensities

        if 'sex' in exp_info_ordered.columns:
            meta['Sex'] = exp_info_ordered['sex'].values
        if 'age' in exp_info_ordered.columns:
            meta['Age'] = exp_info_ordered['age'].values

        csv_path = os.path.join(save_location, f'{csv_base}_metadata.csv')
        meta.to_csv(csv_path, index=False)
        print(f'Metadata exported to: {csv_path}')

    print('Data processing complete.')
    return combined_projection, combined_injection_info, individual_projections, experiment_region_info
